src/derivative_utils.jl

```diff
@@ -371,6 +446,35 @@ function calc_W!(integrator, cache::OrdinaryDiffEqConstantCache, dtgamma, repeat
   W
 end
 
+function calc_W!(nlsolver, integrator, cache::OrdinaryDiffEqConstantCache, dtgamma, repeat_step, W_transform=false)
+  @unpack t,uprev,p,f = integrator
+  @unpack uf = nlsolver
+  mass_matrix = integrator.f.mass_matrix
+  isarray = typeof(uprev) <: AbstractArray
+  # calculate W
+  uf.t = t
+  is_compos = typeof(integrator.alg) <: CompositeAlgorithm
+  if (f isa ODEFunction && islinear(f.f)) || (f isa SplitFunction && islinear(f.f1.f))
+    J = f.f1.f
+    W = WOperator(mass_matrix, dtgamma, J, false; transform=W_transform)
+  elseif DiffEqBase.has_jac(f)
```

Comment:
@devmotion SDIRK tests won't pass until this matches the mutable cache one, which should be a separate PR so that we can get this one correct first.